File: tools/check_policy.mjs
import { loadPolicy } from "@open-policy-agent/opa-wasm";
import fs from "fs";

async function main() {
  // 1. WASM-Datei laden
  const wasmBuffer = fs.readFileSync("../build/policy.wasm");
  const policy = await loadPolicy(wasmBuffer);

  // 2. Event-Payload von Claude Code über stdin einlesen
  const rawInput = fs.readFileSync(0, "utf-8");
  if (!rawInput.trim()) process.exit(0);
  
  const inputData = JSON.parse(rawInput);

  // 3. Rego Policy auswerten
  const results = policy.evaluate(inputData);
  const decision = results[0]?.result;

  // 4. Entscheidung prüfen
  // Falls die Policy fehlschlägt oder allow == false ist:
  if (!decision || decision.allow !== true) {
    const reason = decision?.reason || "Aktion verstößt gegen Compliance-Richtlinien.";
    
    // Fehlerursache an stderr ausgeben (Claude liest dies als Feedback)
    console.error(`[POLICY BLOCKED] ${reason}`);
    
    // Exit Code 2 signalisiert Claude Code, dass die Aktion abgelehnt wurde
    process.exit(2);
  }

  // Exit Code 0 erlaubt die Ausführung
  process.exit(0);
}

main().catch((err) => {
  console.error(`Skriptfehler: ${err.message}`);
  process.exit(2);
});

File: tools/check_policy.py
import sys
import json

from opapywasm import OpaWasmPolicy


def main():
    # 1. WASM-Modul über Wasmtime laden
    policy = OpaWasmPolicy.from_wasm_file("../build/policy.wasm")

    # 2. Payload von Claude Code via stdin lesen
    raw_input = sys.stdin.read()
    if not raw_input.strip():
        sys.exit(0)

    input_data = json.loads(raw_input)

    # 3. Policy via Wasmtime auswerten
    decision = policy.evaluate(input_data)

    # 4. Zugriffsentscheidung auswerten
    if not isinstance(decision, dict) or decision.get("allow") is not True:
        sys.stderr.write("[POLICY BLOCKED] Aktion verstößt gegen Compliance-Richtlinien.\n")
        # Exit-Code 2 signalisiert Claude Code das Abbrechen des Werkzeugaufrufs
        sys.exit(2)

    # Exit-Code 0 erlaubt die Ausführung
    sys.exit(0)

if __name__ == "__main__":
    main()
